pallets/pallet-bonded-coins/src/tests/transactions/refund_account.rs

```diff
@@ -0,0 +1,368 @@
+		});
+}
+
+#[test]
+fn refund_account_works_on_frozen() {
+	let pool_details = generate_pool_details(
+		vec![DEFAULT_BONDED_CURRENCY_ID],
+		get_linear_bonding_curve(),
+		false,
+		Some(PoolStatus::Refunding),
+		Some(ACCOUNT_00),
+		Some(DEFAULT_COLLATERAL_CURRENCY_ID),
+		Some(ACCOUNT_00),
+	);
+	let pool_id = calculate_pool_id(&[DEFAULT_BONDED_CURRENCY_ID]);
+
+	let total_collateral = 10u128.pow(10);
+
+	ExtBuilder::default()
+		.with_pools(vec![(pool_id.clone(), pool_details.clone())])
+		.with_native_balances(vec![(ACCOUNT_01, ONE_HUNDRED_KILT)])
+		.with_collaterals(vec![DEFAULT_COLLATERAL_CURRENCY_ID])
+		.with_bonded_balance(vec![
+			(DEFAULT_COLLATERAL_CURRENCY_ID, pool_id.clone(), total_collateral),
+			(DEFAULT_BONDED_CURRENCY_ID, ACCOUNT_01, total_collateral * 10),
+		])
+		.build()
+		.execute_with(|| {
+			<<Test as crate::Config>::Fungibles as FreezeAccounts<_, _>>::freeze(
+				&DEFAULT_BONDED_CURRENCY_ID,
+				&ACCOUNT_01,
+			)
+			.expect("failed to freeze account prior to testing");
+
+			let origin = RawOrigin::Signed(ACCOUNT_01).into();
+
+			assert_ok!(BondingPallet::refund_account(origin, pool_id.clone(), ACCOUNT_01, 0, 1));
+
+			assert_eq!(
+				<Test as crate::Config>::Fungibles::total_balance(DEFAULT_BONDED_CURRENCY_ID, &ACCOUNT_01),
+				0
+			);
+
+			assert_eq!(
+				<Test as crate::Config>::Fungibles::total_balance(DEFAULT_COLLATERAL_CURRENCY_ID, &ACCOUNT_01),
+				total_collateral
+			);
+
+			// There's only one account, so this should complete the refund
+			System::assert_has_event(Event::<Test>::RefundComplete { id: pool_id }.into());
+		});
+}
+
+#[test]
+fn refund_account_works_with_large_supply() {
+	let currencies = vec![DEFAULT_BONDED_CURRENCY_ID, DEFAULT_BONDED_CURRENCY_ID + 1];
+	let pool_id = calculate_pool_id(&currencies);
+	let pool_details = generate_pool_details(
+		currencies.clone(),
+		get_linear_bonding_curve(),
+		true,
+		Some(PoolStatus::Refunding),
+		Some(ACCOUNT_00),
+		Some(DEFAULT_COLLATERAL_CURRENCY_ID),
+		Some(ACCOUNT_00),
+	);
+
+	let total_collateral = u128::MAX / 2;
+
+	ExtBuilder::default()
+		.with_pools(vec![(pool_id.clone(), pool_details.clone())])
+		.with_native_balances(vec![(ACCOUNT_01, ONE_HUNDRED_KILT)])
+		.with_collaterals(vec![DEFAULT_COLLATERAL_CURRENCY_ID])
+		.with_bonded_balance(vec![
+			(DEFAULT_COLLATERAL_CURRENCY_ID, pool_id.clone(), total_collateral),
+			(currencies[0], ACCOUNT_01, u128::MAX / 3 * 2),
+			(currencies[1], ACCOUNT_01, u128::MAX / 3 * 2),
+		])
+		.build()
+		.execute_with(|| {
+			let origin = RawOrigin::Signed(ACCOUNT_01).into();
+
+			assert_ok!(BondingPallet::refund_account(origin, pool_id.clone(), ACCOUNT_01, 0, 2));
+
+			assert_eq!(
+				<Test as crate::Config>::Fungibles::total_balance(currencies[0], &ACCOUNT_01),
+				0
+			);
+
+			assert_eq!(
+				<Test as crate::Config>::Fungibles::total_balance(DEFAULT_COLLATERAL_CURRENCY_ID, &ACCOUNT_01),
+				total_collateral / 2
+			);
+
```

Comment:
Should we also try to burn the rest of the bonded coins here?

Reply:
sure, doesn't hurt: 33f974ca936b60bd9764fed6735328aec09cf9c7